src/charm.py

```diff
@@ -99,86 +102,142 @@ def _configure_pebble(self) -> None:
+        """The `fiveg_gnb_identity` relation is not mandatory.
 
-    def _get_upf_hostname(self) -> str:
-        """Gets UPF hostname from the `fiveg_n4` relation data bag.
+        If it exists, config files are generated and pushed to the workload.
+        """
+        if not self.model.relations.get(GNB_IDENTITY_RELATION_NAME):
+            logger.warning("Relation %s not available", GNB_IDENTITY_RELATION_NAME)
+            return
+        gnb_existing_content = self._get_existing_config_file(path=GNB_CONFIG_PATH)
+        gnb_config_content = self._get_gnb_config()
+        if not gnb_config_content:
+            logger.warning("gNB config file is invalid")
+            return
+        if not gnb_existing_content or not config_file_content_matches(
+            existing_content=gnb_existing_content, new_content=gnb_config_content
+        ):
+            self._push_gnb_config_file_to_workload(gnb_config_content)
+
+    def _get_existing_config_file(self, path: str) -> str:
+        """Gets the existing config file from the workload.
+
+        Args:
+            path (str): Path to the config file.
 
         Returns:
-            str: UPF hostname
+            str: Content of the config file.
         """
-        fiveg_n4_relation = self.model.get_relation(FIVEG_N4_RELATION_NAME)
-        if not fiveg_n4_relation:
-            return ""
-        if not fiveg_n4_relation.app:
-            raise RuntimeError(
-                f"Application missing from the {FIVEG_N4_RELATION_NAME} relation data"
-            )
-        return fiveg_n4_relation.data[fiveg_n4_relation.app].get("upf_hostname", "")
+        if self._container.exists(path=path):
+            existing_content_stringio = self._container.pull(path=path)
+            return existing_content_stringio.read()
+        return ""
 
-    def _get_upf_port(self) -> Optional[int]:
-        """Gets UPF's N4 port number from the `fiveg_n4` relation data bag.
+    def _get_upf_host_port_list(self) -> List[Tuple[str, int]]:
+        """Gets the list of UPF hosts and ports from the `fiveg_n4` relation data bag.
 
         Returns:
-            int: N4 port number
+            List[Tuple[str, int]]: List of UPF hostnames and ports.
         """
-        fiveg_n4_relation = self.model.get_relation(FIVEG_N4_RELATION_NAME)
-        if not fiveg_n4_relation:
-            return None
-        if not fiveg_n4_relation.app:
-            raise RuntimeError(
-                f"Application missing from the {FIVEG_N4_RELATION_NAME} relation data"
-            )
-        if port := fiveg_n4_relation.data[fiveg_n4_relation.app].get("upf_port", ""):
-            return int(port)
-        return None
+        upf_host_port_list = []
+        for fiveg_n4_relation in self.model.relations.get(FIVEG_N4_RELATION_NAME, []):
+            if not fiveg_n4_relation:
+                logger.warning("Relation %s not available", FIVEG_N4_RELATION_NAME)
+                return []
```

Comment:
Is it even possible to get inside of this condition?

Reply:
no, this legacy code that was needed when using `self.model.relations.get(FIVEG_N4_RELATION_NAME)`, replacing that with `self.model.relations.get(FIVEG_N4_RELATION_NAME, [])` does the job.
Removed.